x-pack/plugins/alerting/server/lib/trim_recovered_alerts.ts

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Logger } from '@kbn/logging';
+import { pick } from 'lodash';
+import { Alert } from '../alert';
+import { AlertInstanceState, AlertInstanceContext } from '../types';
+
+interface TrimmedRecoveredAlertsResult<
+  State extends AlertInstanceState,
+  Context extends AlertInstanceContext,
+  RecoveryActionGroupIds extends string
+> {
+  trimmedAlertsRecovered: Record<string, Alert<State, Context, RecoveryActionGroupIds>>;
+  trimmedAlertsRecoveredCurrent: Record<string, Alert<State, Context, RecoveryActionGroupIds>>;
+}
+
+export function trimRecoveredAlerts<
+  State extends AlertInstanceState,
+  Context extends AlertInstanceContext,
+  RecoveryActionGroupIds extends string
+>(
+  logger: Logger,
+  recoveredAlerts: Record<string, Alert<State, Context, RecoveryActionGroupIds>> = {},
+  currentRecoveredAlerts: Record<string, Alert<State, Context, RecoveryActionGroupIds>> = {},
+  alertLimit: number
+): TrimmedRecoveredAlertsResult<State, Context, RecoveryActionGroupIds> {
+  if (Object.keys(recoveredAlerts).length > alertLimit) {
+    const entries = Object.entries(recoveredAlerts);
+    entries.sort((a, b) => {
```

Comment:
Took at look at this PR before taking off for PTO rest of the year.  Looking good!

One thing about this `sort()` though:

The way this is currently coded we will be calling `getFlappingHistory()` 2 * (n * log2(n)) times (assuming sort is O(n * log2(n)));  n is currently 1000.   

Seems like we should build an array of `{ alertId, flappingLength }`, sort that, and then I guess explicitly delete the alerts from `entries` with `delete entries[alertId]`.  Basically pre-calculate the stuff the body of the sort is calculating, so we don't have to recalculate them during the sort.

I assume however that this will be highly optimized - `getFlappingHistory()` returns the original array, etc, and so probably won't make a noticeable difference.  But imagine we change `getFlappingHistory()` to be more complex, or return a copy, we could then see some performance issues (space and/or time).

Reply:
Oh okay, that makes sense. I will try to optimize this, thanks!